src/VisualStudio/Core/Next/NavigateTo/VisualStudioNavigateToEngineService.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.NavigateTo;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.CodeAnalysis.Remote.Arguments;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Extensions;
+using Microsoft.VisualStudio.LanguageServices.Remote;
+
+namespace Microsoft.VisualStudio.LanguageServices.NavigateTo
+{
+    [ExportWorkspaceService(typeof(INavigateToEngineService), ServiceLayer.Host), Shared]
+    internal class VisualStudioNavigateToEngineService : INavigateToEngineService
+    {
+        public async Task<ImmutableArray<INavigateToSearchResult>> SearchDocumentAsync(
+            Document document, string searchPattern, CancellationToken cancellationToken)
+        {
+            var client = await GetRemoteHostClientAsync(document.Project, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                return await DefaultNavigateToEngineService.SearchDocumentInCurrentProcessAsync(
+                    document, searchPattern, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                return await SearchDocumentInRemoteProcessAsync(
+                    client, document, searchPattern, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<ImmutableArray<INavigateToSearchResult>> SearchDocumentInRemoteProcessAsync(
+            RemoteHostClient client, Document document, string searchPattern, CancellationToken cancellationToken)
+        {
+            var solution = document.Project.Solution;
+
+            using (var session = await client.CreateCodeAnalysisServiceSessionAsync(
+                solution, cancellationToken).ConfigureAwait(false))
+            {
+                var serializableResults = await session.InvokeAsync<SerializableNavigateToSearchResult[]>(
```

Comment:
Navigate to bulk reports results to VS one project at a time.  We could consider being more fine-grained than that in the future.  I think it would be a very nice idea.  But it's not part of the existing implementation, so i didn't want to change anything around that now.